src/lib/protocols/ethereum.c

```diff
@@ -0,0 +1,158 @@
+/*
+ * ethereum.c 
+ *
+ * Copyright (C) 2018-22 - ntop.org
+ *
+ * This file is part of nDPI, an open source deep packet inspection
+ * library based on the OpenDPI and PACE technology by ipoque GmbH
+ *
+ * nDPI is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Lesser General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * nDPI is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with nDPI.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+#include "ndpi_protocol_ids.h"
+#define NDPI_CURRENT_PROTO NDPI_PROTOCOL_ETHEREUM
+#include "ndpi_api.h"
+
+
+/* ************************************************************************** */
+
+static u_int32_t ndpi_ether_make_lru_cache_key(struct ndpi_flow_struct *flow) {
+  u_int32_t key;
+
+  /* network byte order */
+  if(flow->is_ipv6)
+    key = ndpi_quick_hash(flow->c_address.v6, 16) + ndpi_quick_hash(flow->s_address.v6, 16);
+  else
+    key = flow->c_address.v4 + flow->s_address.v4;
+
+  return key;
+}
+
+/* ************************************************************************** */
+
+static void ndpi_ehter_cache_connection(struct ndpi_detection_module_struct *ndpi_struct,
+				 struct ndpi_flow_struct *flow) {
+  if(ndpi_struct->mining_cache)
+    ndpi_lru_add_to_cache(ndpi_struct->mining_cache, ndpi_ether_make_lru_cache_key(flow), NDPI_PROTOCOL_ETHEREUM, ndpi_get_current_time(flow));
+}
+
+/* ************************************************************************** */
+
+static void ndpi_search_ethereum_udp(struct ndpi_detection_module_struct *ndpi_struct,
+				   struct ndpi_flow_struct *flow) {
+  struct ndpi_packet_struct *packet = &ndpi_struct->packet;
+  u_int16_t source = ntohs(packet->udp->source);
+  u_int16_t dest = ntohs(packet->udp->dest);
+
+  NDPI_LOG_DBG(ndpi_struct, "search ETHEREUM UDP\n");
+
+  /* 
+     Ethereum P2P Discovery Protocol
+     https://github.com/ConsenSys/ethereum-dissectors/blob/master/packet-ethereum-disc.c
+  */
+  if((packet->payload_packet_len > 98)
+     && (packet->payload_packet_len < 1280)
+     && ((source == 30303) || (dest == 30303))
```

Comment:
Are you sure that Ethereum uses fixed ports?
The discovery protocol dissector seems not to rely on such.

Reply:
- I'm not sure about this, I took this piece of logic as it's from the mining protocol, I'll double check on this and if it's not true, will change it.

Reply:
- I checked this one and I found that this port is the default port for ethereum discovery protocol.
- [stack overflow answer](https://ethereum.stackexchange.com/questions/809/which-tcp-and-udp-ports-are-required-to-run-an-ethereum-client)
- [ethereum developer](https://ethereum.org/en/developers/docs/networking-layer/network-addresses/#:~:text=58.6%20%2C%20TCP%20port%2030303%20and%20UDP%20discovery%20port%2030301%20.)

- also [here](https://github.com/ethereum/go-ethereum/blob/master/params/bootnodes.go) the same port is used for botnodes ( default nodes that help the client discover other peers when this client's first time to connect to the ethereum network) 
- still from the protocol specification [here](https://github.com/ethereum/devp2p/blob/master/discv4.md) they don't mention anything about used ports 
-  if we removed it, we only left with lengths which is not a strong indication. 
- this was there (as part of the mining protocol) and no complaints about it yet.
- we can have a ticket for later to improve the detection and find other strong patterns, and we can then remove the port.